### PR TITLE
Add context to negotiate requests

### PR DESCRIPTION
**Pull request by vicancy**
For ASP.NET one, add `IOwinContext` to `GetNegotiateEndpoint` for `IEndpointRouter`, for core one, add `HttpContext`

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/EndpointRouters/DefaultEndpointRouter.cs [hunk 1]
```diff
@@ -3,24 +3,13 @@
 
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.AspNetCore.Http;
 using Microsoft.Azure.SignalR.Common;
 
 namespace Microsoft.Azure.SignalR
 {
-    internal class DefaultEndpointRouter : IEndpointRouter
+    internal class DefaultMessageRouter : IMessageRouter
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/EndpointRouters/DefaultEndpointRouter.cs]: Rename file name as well.

File: src/Microsoft.Azure.SignalR/EndpointRouters/IEndpointRouter.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.Azure.SignalR
+{
+    public interface IEndpointRouter : IMessageRouter
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR/EndpointRouters/IEndpointRouter.cs]: what about `INegotiateRouter` and `IMessageRouter`?

File: src/Microsoft.Azure.SignalR.AspNet/Middleware/NegotiateMiddleware.cs [hunk 3]
```diff
@@ -99,7 +99,20 @@ private Task ProcessNegotiationRequest(IOwinContext owinContext, HostContext con
             IServiceEndpointProvider provider;
             try
             {
-                provider = _endpointManager.GetEndpointProvider(_router.GetNegotiateEndpoint(_endpointManager.Endpoints));
+                provider = _endpointManager.GetEndpointProvider(_router.GetNegotiateEndpoint(owinContext, _endpointManager.Endpoints));
+
+                // Consider it as internal server error when we don't successfully get negotiate response
+                if (provider == null)
```
Comment (chenkennt) [inline src/Microsoft.Azure.SignalR.AspNet/Middleware/NegotiateMiddleware.cs]: should check context first? if it's not 200 then return

File: src/Microsoft.Azure.SignalR/EndpointRouters/IEndpointRouter.cs (hunk 2 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR/EndpointRouters/IEndpointRouter.cs]: Tend to not make that big change

File: src/Microsoft.Azure.SignalR.AspNet/Middleware/NegotiateMiddleware.cs [hunk 4]
```diff
@@ -99,7 +99,26 @@ private Task ProcessNegotiationRequest(IOwinContext owinContext, HostContext con
             IServiceEndpointProvider provider;
             try
             {
-                provider = _endpointManager.GetEndpointProvider(_router.GetNegotiateEndpoint(_endpointManager.Endpoints));
+                provider = _endpointManager.GetEndpointProvider(_router.GetNegotiateEndpoint(owinContext, _endpointManager.Endpoints));
+
+                if (context.Response.StatusCode >= 400)
```
Comment (chenkennt) [inline src/Microsoft.Azure.SignalR.AspNet/Middleware/NegotiateMiddleware.cs]: is != 200 better?

File: src/Microsoft.Azure.SignalR.Common/EndpointRouters/DefaultMessageRouter.cs [hunk 5]
```diff
@@ -55,7 +44,8 @@ public IEnumerable<ServiceEndpoint> GetEndpointsForGroup(string groupName, IEnum
         }
 
         /// <summary>
-        /// Broadcast to all available endpoints
+        /// Broadcast to all available endpoints, note that this one is only called when the SDK is not able to identify where the connectionId is.
+        /// When the outcoming connectionId happens to be also connected to this app server, SDK can directly send the messages back to that connectionId
```
Comment (chenkennt) [inline src/Microsoft.Azure.SignalR.Common/EndpointRouters/DefaultMessageRouter.cs]: You should put this comment to the interface

**Pull request merged**